### ISSUE TITLE
False-positive `missing-argument` when setting method from staticmethod without decorator

### ISSUE THREAD

**Issue by augustfe** [astral-sh/ty#2724]
### Summary

ty reports `error[missing-argument]` when setting a method in a class from `staticmethod(foo)`. It works as expected when `__call__` is set with the `@staticmethod` decorator. I discovered this behaviour when working with the SymPy v1.14 SingletonRegistry.

Minimal reproduction:
```python
def foo(*args, **kwargs) -> None:
    print("foo", args, kwargs)

class A:
    __call__ = staticmethod(foo)
    bar = staticmethod(foo)

a = A()
a()
a.bar()
a(5)
a.bar(5)
a(x=10)
a.bar(x=10)
```
[Playground link](https://play.ty.dev/2b99e521-d74d-401d-ba1e-74c6b8daed32)

Result from `ty check`:
```system
❯ uv run ty check test.py                                                                                             13:23 󰴈 
error[missing-argument]: No arguments provided for required parameters `*args`, `**kwargs`
  --> test.py:17:1
   |
16 | a = A()
17 | a()
   | ^^^
18 | a.bar()
   |
info: Union variant `[**_P'return, _R_co'return](**_P'return@staticmethod) -> _R_co'return` is incompatible with this call site
info: Attempted to call union type `A`
info: These arguments are required because `ParamSpec` `_P'return` could represent any set of parameters at runtime
info: rule `missing-argument` is enabled by default

error[missing-argument]: No arguments provided for required parameters `*args`, `**kwargs`
  --> test.py:18:1
  ...

error[missing-argument]: No argument provided for required parameter `**kwargs`
  --> test.py:20:1
  ...
error[missing-argument]: No argument provided for required parameter `**kwargs`
  --> test.py:21:1
  ...
error[missing-argument]: No argument provided for required parameter `*args`
  --> test.py:22:1
  ...
error[missing-argument]: No argument provided for required parameter `*args`
  --> test.py:23:1
  ...

Found 6 diagnostics
```

Actual behaviour:
```system
$ uv run test.py
foo () {}
foo () {}
foo (5,) {}
foo (5,) {}
foo () {'x': 10}
foo () {'x': 10}
```

### Version

ty 0.0.15

**Comment by sharkdp** [comment]
Thank you for reporting this. Here is a smaller example:
```py
def foo(*args, **kwargs) -> None:
    print("foo", args, kwargs)


class A:
    bar = staticmethod(foo)


A.bar()
```

The full error message is
```
error[missing-argument]: No arguments provided for required parameters `*args`, `**kwargs`
 --> /home/shark/playground/test.py:9:1
  |
9 | A.bar()
  | ^^^^^^^
  |
info: Union variant `[**_P'return, _R_co'return](**_P'return@staticmethod) -> _R_co'return` is incompatible with this call site
info: Attempted to call union type `Unknown | ([**_P'return, _R_co'return](**_P'return@staticmethod) -> _R_co'return)`
info: These arguments are required because `ParamSpec` `_P'return` could represent any set of parameters at runtime
info: rule `missing-argument` is enabled by default
```


Looks like this might be related to https://github.com/astral-sh/ruff/pull/22820 (cc @dhruvmanila)

**Comment by dhruvmanila** [comment]
Yeah, this seems like a bug but it's probably not related to https://github.com/astral-sh/ruff/pull/22820.

Currently, the revealed type for `A.bar` is:
```py
Unknown | ([**_P'return, _R_co'return](**_P'return@staticmethod) -> _R_co'return)
```

while, it should've been:
```py
Unknown | (...) -> None
```

given that the signature of `foo` is `def foo(...) -> None`.

The type variables aren't being specialized in this case.

I think this is because we don't have proper support for `staticmethod` because a custom method works fine:

```py
from typing import Callable

def foo(*args, **kwargs) -> None:
    print("foo", args, kwargs)

def custom[**P, R](fn: Callable[P, R]) -> Callable[P, R]:
    return fn

class A:
    bar = staticmethod(foo)
    baz = custom(foo)


# Revealed type: `Unknown | ([**_P'return, _R_co'return](**_P'return@staticmethod) -> _R_co'return)`
reveal_type(A.bar)

# Revealed type: `Unknown | ((...) -> None)`
reveal_type(A.baz)
```

Playground: https://play.ty.dev/ba6b5106-d82b-4277-a07e-a7c8f66333aa

**Comment by SebbyLaw** [comment]
another repro: https://play.ty.dev/8e647d4f-a7c4-450a-99b3-491c3c5a2690
```py
class One:
    @staticmethod
    def foo(bar: int) -> str:
        return str(bar)


class Two:
    foo = staticmethod(One.foo)


two = Two()
print(two.foo(42))
```

**Comment by augustfe** [comment]
Might be a separate issue, but ty does not raise any errors for this, even though a TypeError is thrown.

```python
class One:
    @staticmethod
    def foo(bar: int) -> str:
        return str(bar)

class Two:
    foo = One.foo

two = Two()
print(two.foo(42))  # TypeError: One.foo() takes 1 positional argument but 2 were given
```
https://play.ty.dev/66be446a-3805-4b66-8735-d5e40f72bc4d

**Comment by carljm** [comment]
The last example is unrelated (and all other type checkers have the same behavior as ty here). It's another case of the basic problem that type checkers have no way to determine whether a `Callable` type will be inhabited by an object that does or doesn't implement the bound-method descriptor protocol. `staticmethod.__get__` in typeshed just returns a normal Callable type, so type checkers just have to guess. See #491 for more discussion.

**Comment by carljm** [comment]
The `'return` in the name of those typevars is suspicious. It looks like maybe we are considering `staticmethod.__get__` to be returning a standalone generic callable type? Which is wrong, because those typevars are in the generic context of `staticmethod`. cc @dcreager

**Comment by Hugo-Polloli** [comment]
I can reproduce on `dc65d12668^` (with the `'return` `ParamSpec` / `missing-argument` errors), but not on `dc65d12668` or any later commit, so this seems to have been fixed by [#21902](https://github.com/astral-sh/ruff/pull/21902)

Tested using this mdtest locally, I'm not sure this is worth opening a PR over
<img width="1334" height="932" alt="Image" src="https://github.com/user-attachments/assets/50b2cc1b-840f-4ca8-ae62-9eaafcd75d79" />

**Comment by carljm** [comment]
@Hugo-Polloli I think it's worth a PR to add an mdtest for this case and avoid regression!
